src/ripple/ledger/impl/ApplyStateTable.cpp

```diff
@@ -0,0 +1,589 @@
+ApplyStateTable::peek (ReadView const& base,
+    Keylet const& k)
+{
+    auto iter = items_.lower_bound(k.key);
+    if (iter == items_.end() ||
+        iter->first != k.key)
+    {
+        auto const sle = base.read(k);
+        if (! sle)
+            return nullptr;
+        // Make our own copy
+        using namespace std;
+        iter = items_.emplace_hint (iter,
+            piecewise_construct,
+                forward_as_tuple(sle->key()),
+                    forward_as_tuple(Action::cache,
+                        make_shared<SLE>(*sle)));
+        return iter->second.second;
+    }
+    auto const& item = iter->second;
+    auto const& sle = item.second;
+    switch (item.first)
+    {
+    case Action::erase:
+        return nullptr;
+    case Action::cache:
+    case Action::insert:
+    case Action::modify:
+        break;
+    };
+    if (! k.check(*sle))
+        return nullptr;
+    return sle;
+}
+
+void
+ApplyStateTable::erase(
+    ReadView const& base,
+        std::shared_ptr<SLE> const& sle)
+{
+    auto const iter =
+        items_.find(sle->key());
+    if (iter == items_.end())
+        LogicError("ApplyStateTable::erase: missing key");
+    auto& item = iter->second;
+    if (item.second != sle)
+        LogicError("ApplyStateTable::erase: unknown SLE");
+    switch(item.first)
+    {
+    case Action::erase:
+        LogicError("ApplyStateTable::erase: double erase");
+        break;
+    case Action::insert:
+        items_.erase(iter);
+        break;
+    case Action::cache:
+    case Action::modify:
+        item.first = Action::erase;
+        break;
+    }
+}
+
+void
+ApplyStateTable::rawErase (ReadView const& base,
+    std::shared_ptr<SLE> const& sle)
+{
+    using namespace std;
+    auto const result = items_.emplace(
+        piecewise_construct,
+            forward_as_tuple(sle->key()),
+                forward_as_tuple(Action::erase, sle));
+    if (result.second)
+        return;
+    auto& item = result.first->second;
+    switch(item.first)
+    {
+    case Action::erase:
+        LogicError("ApplyStateTable::rawErase: double erase");
+        break;
+    case Action::insert:
+        items_.erase(result.first);
+        break;
+    case Action::cache:
+    case Action::modify:
+        item.first = Action::erase;
+        item.second = sle;
+        break;
+    }
+}
+
+void
+ApplyStateTable::insert (ReadView const& base,
+    std::shared_ptr<SLE> const& sle)
+{
+    auto const iter =
+        items_.lower_bound(sle->key());
+    if (iter == items_.end() ||
+        iter->first != sle->key())
+    {
+        using namespace std;
+        items_.emplace_hint(iter,
+            piecewise_construct,
+                forward_as_tuple(sle->key()),
+                    forward_as_tuple(Action::insert, sle));
+        return;
+    }
+    auto& item = iter->second;
+    switch(item.first)
+    {
+    case Action::cache:
+        LogicError("ApplyStateTable::insert: already cached");
```

Comment:
All that the log would do is make us want to look at the core dump. But I agree that the error should be logged. If you have a core dump, you can at least tell what to do with it by looking at the end of the log. If not, at least you have something to go on.